server/src/main/java/org/apache/iotdb/db/metadata/schemaregion/SchemaRegionSchemaFileImpl.java

```diff
@@ -297,20 +349,26 @@ private int initFromLog() throws IOException {
     }
   }
 
-  private int applyMLog(SchemaLogReader<ISchemaRegionPlan> mLogReader) {
-    int idx = 0;
+  /**
+   * Redo metadata log file.
+   *
+   * @param startIndex start index to redo MLog
+   * @return next index of the logFile
+   */
+  private long applyMLog(SchemaLogReader<ISchemaRegionPlan> mLogReader, long startIndex) {
+    long nextIndex = 0;
     ISchemaRegionPlan plan;
     RecoverPlanOperator recoverPlanOperator = new RecoverPlanOperator();
     RecoverOperationResult operationResult;
     while (mLogReader.hasNext()) {
       try {
         plan = mLogReader.next();
-        idx++;
+        nextIndex++;
       } catch (Exception e) {
         logger.error("Parse mlog error at lineNumber {} because:", idx, e);
         break;
       }
-      if (plan == null) {
+      if (nextIndex <= startIndex || plan == null) {
         continue;
       }
```

Comment:
The logs before checkpoint will not be redo but still read from disk. What about recording the position/offset of checkpoint in mlog? The read operation of logs before checkpoint will be saved. 

If this takes effort to implement such mechanism. What about we archive the executed logs into another log file, like ```mlog_archive_xxx```, and they won't be read or executed in next recover process, since if we successfully recover, all the schema has already been stored in schema file.

Reply:
I use position as checkpoint now. Recovery time has been reduced from 3000ms to 1300ms.